### DISCUSSION TITLE
On '@' and reserved words

### DISCUSSION THREAD

**Discussion by stephentoub** [dotnet/csharplang#446] [General]
_From @juliusfriedman on July 19, 2016 16:28_

Attempting to compile the following code:

```
static void @_(object someName)
        {
            object @someName = null;

            return;
        }
```

results in a compilation error:

> Error 79  A local variable named 'someName' cannot be declared in this scope because it would give a different meaning to 'someName', which is already used in a 'parent or current' scope to denote something else.

Additionally it seems that the `@` character is used as an `Escape` in some regard in the parser logic and thus it's appearance causes the character which follows to lose it's meaning as can be seen in the following example:

```
 static void @(object someName)
        {
            object @someName = null;

            return;
        }
```

Thanks for your time and consideration.

_Copied from original issue: dotnet/corefx#10152_

**Comment by svick** [comment]
This is the expected behavior. `@` is meant to be used to allow you to use keywords as identifiers (e.g. if you want to create a class with name `class`, you have to write `class @class`). It's also possible to use it with normal identifiers, but then the `@` is not considered part of the identifier.

This explains the behavior you see:

* `someName` and `@someName` are considered to be the same identifier, which is why you get an error when you try to use them as two different identifiers.
* `@` on its own is not a valid identifier, which is why the compiler is very confused if you try to declare a method with just it as its name and throws up a lot of parsing errors. (That behavior could be improved, but that's just a matter of getting nicer error messages, not a bug in the compiler.)

If you want to know more, all this is explained in [the *Identifiers* section of the spec](https://github.com/dotnet/csharplang/blob/master/spec/lexical-structure.md#identifiers).
